### PR TITLE
React router hash link

### PR DESCRIPTION
**Pull request by jackahl**
potential fix for #4164

### CODE REVIEW

**Review by tiberiuichim — APPROVED** [review]

**Review by ksuess — APPROVED** [review]
Even scrolling with offset works fine.
Just on pages with an accordion, the scrolling does not touch exactly the offset (default or with scroll function).

+1 from me.

**Comment by ksuess** [comment]
The ScrollToTop still interferes. 
I would like to suggest to modify the default "scroll to top" behavior by adding a condition "only if no hash".
https://github.com/plone/volto/blob/abe52c2881603f6387b0a5c9de19c876ae52d8ea/src/helpers/ScrollToTop/ScrollToTop.jsx#L29-L33

```
  componentDidUpdate(prevProps) {
    if (
      this.props.location?.pathname !== prevProps.location?.pathname &&
      !this.props.location?.hash
    ) {
      window.scrollTo(0, 0);
    }
  }
```

**Comment by sneridagh** [comment]
During the Volto Team meeting we decided to:
- adopt the library into Volto itself
- take advantage of the above and expose as Volto settings some of the library settings
- Fine tuning the whole setup (ScrollToTop)

@jackahl , @ksuess can you coordinate the efforts? Thanks!

**Comment by jackahl** [comment]
> During the Volto Team meeting we decided to:
> 
>     * adopt the library into Volto itself
> 
>     * take advantage of the above and expose as Volto settings some of the library settings
> 
>     * Fine tuning the whole setup (ScrollToTop)
> 
> 
> @jackahl , @ksuess can you coordinate the efforts? Thanks!

sure. Pushed the code for @ksuess proposal. 
@sneridagh what features of the library do you want to expose via the registry? The smoothscroll?

**Comment by sneridagh** [comment]
@jackahl yes, and she mentioned another one I think.

**Comment by jackahl** [comment]
@sneridagh in the libraries documentation (https://github.com/rafgraph/react-router-hash-link#readme) I could not see any other setting, that would make sense to put into the registry, except for the smooth scroll. Did I miss something?

**Comment by sneridagh** [comment]
Let's wait for @ksuess to weigh in. I'll look into the current shape! Thanks!

**Comment by ksuess** [comment]
Originally I thought to export a scrollWithOffset function. But now I think the use case is too rare.

What about providing an example in the documentation:

````
Scrolling with offset:

```js
const scrollWithOffset = (el, offset = null) => {
  let yCoordinate = el.getBoundingClientRect().top + window.pageYOffset;
  const yOffset = offset ? -offset : 0;
  window.scrollTo(0, yCoordinate + yOffset, { behavior: 'smooth' });
};
```

```html
<ConditionalLink
  to={`${el.pageurl}#${el.id}`}
  condition={!isEditMode}
  scroll={(el) => scrollWithOffset(el, 25)}
>
  {el.text}
</ConditionalLink>
```
````

**Comment by jackahl** [comment]
@ksuess I agree with it probably being a rather rare usecase. If we include a registry option for every edgecase we might bloat the registry unnecessarily and make it harder to use. And I just had the same thought about our documentation. If we include too many edgecases there it might become harder to use as well. Especially if its documented somewhere else (the packages documentation) anyways. Maybe a reference to the packages documentation somewhere in our docs might be enough? Though I am not sure how much of an 'edgecase' offset scrolling is. Thats something that more experienced devs than me are probably better in determining.

**Comment by ksuess** [comment]
> @ksuess I agree with it probably being a rather rare usecase. If we include a registry option for every edgecase we might bloat the registry unnecessarily and make it harder to use. And I just had the same thought about our documentation. If we include too many edgecases there it might become harder to use as well. Especially if its documented somewhere else (the packages documentation) anyways. Maybe a reference to the packages documentation somewhere in our docs might be enough? Though I am not sure how much of an 'edgecase' offset scrolling is. Thats something that more experienced devs than me are probably better in determining.

Yes, sure, I'm fine with not including the offset example.
If you integrate the code of the package in Volto, you will mention the package. Who has the need for an offset can check its README and find the link to the tip.

**Review by ksuess — APPROVED** [review]

**Comment by tiberiuichim** [comment]
@sneridagh I'm fine with the merge now.

**Comment by ksuess** [comment]
I'm fine with merging now with the dependency.

**Review by sneridagh — APPROVED** [review]

File: src/helpers/ScrollToTop/ScrollToTop.jsx [hunk 1]
```diff
@@ -27,7 +27,10 @@ class ScrollToTop extends React.Component {
    * @memberof ScrollToTop
    */
   componentDidUpdate(prevProps) {
-    if (this.props.location?.pathname !== prevProps.location?.pathname) {
+    if (
+      !this.props.location?.pathname.hash &&
```
Comment (ksuess) [inline src/helpers/ScrollToTop/ScrollToTop.jsx:31]: Typo. This should read
!this.props.location?.hash &&

**Review by ksuess — COMMENTED** [review]
@jackahl, can you please fix the condition on hash.

File: src/helpers/ScrollToTop/ScrollToTop.jsx (hunk 1 above)
Comment (nileshgulia1) [reply src/helpers/ScrollToTop/ScrollToTop.jsx:31]: I took care of this in https://github.com/plone/volto/pull/4287/files#diff-1712f4ab0dc421cacb610886108bb1e5cd7978212604901774cf341a5fcf9ed1R30

Comment (nileshgulia1) [reply src/helpers/ScrollToTop/ScrollToTop.jsx:31]: Also, we need to account for "copy link to highlight" feature on browsers. Right now it causes problems.

**Pull request merged**